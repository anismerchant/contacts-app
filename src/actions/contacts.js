import database from '../firebase/firebase'

// Create
export const addContact = (contact) => ({
    type: 'ADD_CONTACT',
    contact
});

export const beginAddContact = ({firstname = '', lastname = '', email = '', cell = ''} = {}) => {
    return (dispatch) => {
        const contact = { firstname, lastname, email, cell }

        return database.ref('contacts').push(contact).then( (ref) => {
            dispatch(addContact({
                id: ref.key,
                ...contact
            }));
        });
    };
};

// Read 
export const setContacts = (contacts) => ({
    type: 'SET_CONTACTS',
    contacts
});

export const beginSetContacts = () => {
    return (dispatch) => { 
        return database.ref('contacts').once('value').then( (snapshot) => {
            const contacts = [];

            snapshot.forEach((childSnapshot) => {
                contacts.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });
            dispatch(setContacts(contacts));
        });
    };
};

// Update
export const editContact = (id, updates) => ({
    type: 'EDIT_CONTACT',
    id,
    updates
});

export const beginEditContact = (id, updates) => {
    return (dispatch) => {
        return database.ref(`contacts/${id}`).update(updates).then( () => {
            dispatch(editContact(id, updates))
        })
    }
}

// Delete
export const removeContact = ({id} = {}) => ({
    type: 'REMOVE_CONTACT',
    id
});

export const beginRemoveContact = ({id} = {}) => {
    return (dispatch) => {
        return database.ref(`contacts/${id}`).remove().then( () => {
            dispatch(removeContact({ id }));
        })
     
    };
};
