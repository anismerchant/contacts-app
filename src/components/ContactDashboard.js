import React from 'react';

const ContactDashboard = () => (
    <div>
        Contact Dashboard
    </div>
);

export default ContactDashboard;