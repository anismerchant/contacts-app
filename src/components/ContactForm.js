import React, { Component } from 'react';

class ContactForm extends Component {
    render() {
        return (
            <form>
                <input 
                    type="text"
                    placeholder="First Name"
                />
                <input 
                    type="text"
                    placeholder="Last Name"
                />
                <input 
                    type="email"
                    placeholder="Email" 
                    id="email"
                    pattern=".+@globex.com" 
                    size="30" 
                />
                <input 
                    type="tel"
                    placeholder="Cell" 
                    id="cell" 
                    name="cell"
                    pattern="[0-9]{3}-[0-9]{3}-[0-9]{4}"
                />
            </form>
        );
    }
}

export default ContactForm;