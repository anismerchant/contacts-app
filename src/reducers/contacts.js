const contactsReducerDefaultState = [];

export const contactsReducer = ( state = contactsReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_CONTACT':
            return [
                ...state,
                action.contact
            ];
        case 'SET_CONTACTS':
            return action.contacts;
        case 'EDIT_CONTACT':
            return state.map( (contact) => {
                return (contact.id === action.id) 
                ? {...contact, ...action.updates} 
                : contact
            });
        case 'REMOVE_CONTACT':
            return state.filter( ({id}) => id !== action.id );
        default:
            return state;
    }
}
