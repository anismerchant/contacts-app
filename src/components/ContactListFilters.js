import React from 'react';
import { connect } from 'react-redux';
import { setNameFilter, sortByFirstName, sortByLastName } from '../actions';

const ContactListFilters = (props) => (
    <div className="contactlist--filters">
        <input 
            type="text" 
            placeholder="Search"
            value={props.filters.name} 
            onChange={(e) => props.dispatch(setNameFilter(e.target.value))}
        />
        <select
            value={props.filters.sortBy}
            onChange = {(e) => 
            {
                if (e.target.value === "firstname") {
                    props.dispatch(sortByFirstName())
                } else if (e.target.value === "lastname") {
                    props.dispatch(sortByLastName())
                } 
            }}
        >   
            <option className="contactlist--sortby" value="" disabled>Sort by</option>
            <option className="contactlist--firstname" value="firstname">First Name</option>
            <option className="contactlist--lastname" value="lastname">Last Name</option>
        </select>
    </div>

);

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    };
};

export default connect(mapStateToProps)(ContactListFilters);
