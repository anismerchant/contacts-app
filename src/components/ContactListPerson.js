import React from 'react';
import { removeContact } from '../actions/contacts';
import { connect } from 'react-redux';

// de-structured props
const ContactListPerson = ({dispatch, id, firstname, lastname, email, cell}) => (
    <div>
        <p>{`${firstname} ${lastname} ${email} ${cell}`}</p>
        <button onClick={ () => {dispatch(removeContact({id}))
        }}>Remove</button>
    </div>
);

export default connect()(ContactListPerson);