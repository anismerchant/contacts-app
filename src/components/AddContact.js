import React from 'react';
import { connect } from 'react-redux';
import ContactForm from './ContactForm';
import { beginAddContact }  from '../actions/contacts'

export const AddContact = (props) => {
    return (
        <div>
            <h3>Add Contact</h3>
            <ContactForm 
                onSubmit={ (contact) => {
                    props.dispatch(beginAddContact(contact));
                    props.history.push('/dashboard');
                }}
            />
        </div>
    );
}

export default connect()(AddContact);