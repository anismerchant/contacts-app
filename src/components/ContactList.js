import React from 'react';
import { connect } from 'react-redux';
import ContactListPerson from './ContactListPerson';
import findContacts from '../selectors/findContacts';

// de-structured props
const ContactList = ({contacts}) => (
    <div>
        {contacts.length}
        {contacts.map((contact) => <ContactListPerson key={contact.id} {...contact} />)}
    </div>
);

const mapStateToProps = (state) => {
    return {
        contacts: findContacts(state.contacts, state.filters)
    }
}

export default connect(mapStateToProps)(ContactList);