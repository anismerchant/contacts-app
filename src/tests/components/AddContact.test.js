import React from 'react';
import { shallow } from 'enzyme';
import { AddContact } from '../../components/AddContact';
import contacts from '../seed/contacts';

let beginAddContact, history, wrapper;

beforeEach( () => {
    beginAddContact = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(<AddContact beginAddContact={beginAddContact} history = {history} />)
});

test('This should correctly render AddContact component.' , () => {
    expect(wrapper).toMatchSnapshot();
});