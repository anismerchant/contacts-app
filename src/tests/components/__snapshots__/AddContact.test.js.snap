// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`This should correctly render AddContact component correnctly. 1`] = `
<div>
  <h3>
    Add Contact
  </h3>
  <ContactForm
    onSubmit={[Function]}
  />
</div>
`;
